### PR TITLE
Get provider link [POL-34]

### PR DESCRIPTION
**Pull request by snwessel**
Creating an endpoint to get a provider link. This includes:
- Getting the user id from SAM
- Setting up error handling 
- Figuring out a process for reading from and writing to the DB (and testing that process)

### CODE REVIEW

File: service/src/main/java/bio/terra/externalcreds/services/SamService.java [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+  public String getUserIdFromSam() {
+    try {
+      String header = request.getHeader("authorization");
+      if (header == null) throw new UnauthorizedException("User is not authorized");
+      String accessToken = BearerTokenParser.parse(header);
+
+      return samUsersApi(accessToken).getUserStatusInfo().getUserSubjectId();
+    } catch (ApiException e) {
+      throw new ExternalCredsException(e, HttpStatus.INTERNAL_SERVER_ERROR);
```
Comment (dvoet) [inline service/src/main/java/bio/terra/externalcreds/services/SamService.java]: we should handle a 404 response from sam differently. If sam returns a 404, I would be inclined to return a 403 from ECM but I am happy to hear other opinions.

File: service/src/main/java/bio/terra/externalcreds/services/SamService.java [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+public class SamService {
+
+  private final HttpServletRequest request;
+
+  @Value("${externalcreds.sam-address}")
+  private String samBasePath;
+
+  public SamService(HttpServletRequest request) {
+    this.request = request;
+  }
```
Comment (dvoet) [inline service/src/main/java/bio/terra/externalcreds/services/SamService.java]: can we confirm that a new SamService instance is created for each request?

File: service/src/test/java/bio/terra/externalcreds/service/LinkedAccountServiceTest.java [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.Test;
+import org.mockito.Mockito;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.test.mock.mockito.MockBean;
+
+public class LinkedAccountServiceTest extends BaseTest {
+  @Autowired private LinkedAccountService linkedAccountService;
+
+  @MockBean private LinkedAccountDAO linkedAccountDAO;
```
Comment (dvoet) [inline service/src/test/java/bio/terra/externalcreds/service/LinkedAccountServiceTest.java]: I don't think we should use a mock here, use the real thing. In this particular test it does not really matter but future tests may rely on complex database state making for a complicated and brittle mock.

File: service/src/test/java/bio/terra/externalcreds/controllers/OidcApiControllerTest.java [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+public class OidcApiControllerTest extends BaseTest {
+
+  @Autowired private MockMvc mvc;
+
+  @MockBean private LinkedAccountService linkedAccountService;
+  @MockBean private SamService samService;
+
+  @Test
+  void testGetLink() throws Exception {
```
Comment (dvoet) [inline service/src/test/java/bio/terra/externalcreds/controllers/OidcApiControllerTest.java]: also test 404 response and when sam returns 404

**Review by dvoet — CHANGES_REQUESTED** [review]

**Review by dvoet — APPROVED** [review]

File: service/src/main/java/bio/terra/externalcreds/controllers/OidcApiController.java [hunk 5]
```diff
@@ -27,23 +27,23 @@
...
-    this.samService = samService;
+      LinkedAccountService linkedAccountService,
+      ObjectMapper mapper,
+      ProviderService providerService,
+      SamService samService) {
     this.request = request;
+    this.linkedAccountService = linkedAccountService;
     this.mapper = mapper;
+    this.providerService = providerService;
+    this.samService = samService;
```
Comment (dvoet) [inline service/src/main/java/bio/terra/externalcreds/controllers/OidcApiController.java]: why this change, specifically the reordering?

Comment (snwessel) [reply service/src/main/java/bio/terra/externalcreds/controllers/OidcApiController.java]: Honestly, I think it may have been accidental. We had been walking through the PR as a group and @zarsky-broad  suggested making these alphabetical, but then we decided it wasn't worth it and went to change it back (but must not have undone all of our changes, hence the new order).

I'd be happy to change it back or make it alphabetical, but I'm not sure if they were in any particular order to begin with.

Comment (dvoet) [reply service/src/main/java/bio/terra/externalcreds/controllers/OidcApiController.java]: no need to revert, I just want to make sure there is no expectation that our parameter lists are alphabetical
